File: android/ssabree/app/src/main/java/com/ssafy/ssabree/features/login/viewmodel/LoginViewModel.kt
package com.ssafy.ssabree.features.login.viewmodel

import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.ssafy.ssabree.core.repository.AuthRepository
import kotlinx.coroutines.launch

private const val TAG = "LoginViewModel"

data class LoginUiState(
    val email: String = "",
    val password: String = "",
    val isLoading: Boolean = false,
    val errorMessage: String? = null,
    val isLoginSuccess: Boolean = false
)

class LoginViewModel(
    private val authRepository: AuthRepository
) : ViewModel() {

    var uiState by mutableStateOf(LoginUiState())
        private set

    fun onEmailChange(newEmail: String) {
        uiState = uiState.copy(
            email = newEmail,
            errorMessage = null
        )
    }

    fun onPasswordChange(newPw: String) {
        uiState = uiState.copy(
            password = newPw,
            errorMessage = null
        )
    }

    fun login(onLoginSuccess: () -> Unit) {
        val email = uiState.email.trim()
        val pw = uiState.password

        if (email.isBlank() || pw.isBlank()) {
            return
        }

        viewModelScope.launch {
            uiState = uiState.copy(
                isLoading = true,
                errorMessage = null
            )

            val result = authRepository.login(email, pw)

            result
                .onSuccess {
                    Log.d(TAG, "login: succeed")
                    // AuthRepositoryImpl에서 이미 토큰 저장, FCM 동기화까지 수행함
                    uiState = uiState.copy(
                        isLoading = false,
                        isLoginSuccess = true
                    )
                    onLoginSuccess()
                }
                .onFailure { e ->
                    Log.d(TAG, "login: failed (${e.message})")
                    uiState = uiState.copy(
                        isLoading = false,
                        errorMessage = e.message
                    )
                }
        }
    }
}

File: iOS/ssabree/ssabree/Features/Login/LoginViewModel.swift
import Foundation
import Observation

struct LoginUiState {
    var email: String = ""
    var password: String = ""
    var isLoading: Bool = false
    var errorMessage: String?
    var isLoginSuccess: Bool = false
}

@MainActor
@Observable
final class LoginViewModel {
    private(set) var uiState = LoginUiState()

    private let authRepository: AuthRepository

    init(authRepository: AuthRepository) {
        self.authRepository = authRepository
    }

    func onEmailChange(_ newEmail: String) {
        uiState.email = newEmail
        uiState.errorMessage = nil
    }

    func onPasswordChange(_ newPassword: String) {
        uiState.password = newPassword
        uiState.errorMessage = nil
    }

    func login(onLoginSuccess: @escaping () -> Void) {
        let email = uiState.email.trimmingCharacters(in: .whitespacesAndNewlines)
        let password = uiState.password

        guard !email.isEmpty, !password.isEmpty else {
            uiState.errorMessage = "이메일과 비밀번호를 입력해주세요"
            return
        }

        Task {
            uiState.isLoading = true
            uiState.errorMessage = nil

            let result = await authRepository.login(email: email, password: password)

            switch result {
            case .success:
                print("Login succeeded")
                uiState.isLoading = false
                uiState.isLoginSuccess = true

                // 로그인 성공 시 FCM 토큰 서버에 동기화
                FcmTokenSyncer.shared.syncIfAuthenticated()

                onLoginSuccess()

            case .failure(let error):
                print("Login failed: \(error.localizedDescription)")
                uiState.isLoading = false
                uiState.errorMessage = error.localizedDescription
            }
        }
    }

    func clearError() {
        uiState.errorMessage = nil
    }
}
